packages/frontend/src/pages/TokenSwap/ui/Success.js

```diff
@@ -5,78 +5,100 @@ import styled from 'styled-components';
+    > button {
+        display: block;
+        width: 100%;
     }
 `;
 
-export default function Success({
-    tokenIn,
-    amountIn,
-    tokenOut,
-    amountOut,
-    transactionHash,
-    onClickContinue,
-    onClickGoToExplorer,
-}) {
-    const messageData = {
-        amountFrom: `${amountIn} ${tokenIn?.onChainFTMetadata?.symbol}`,
-        amountTo: `${amountOut} ${tokenOut?.onChainFTMetadata?.symbol}`,
+export default function Success({ onClickContinue, onClickGoToExplorer }) {
+    const {
+        swapState: { tokenIn, amountIn, tokenOut, amountOut, lastSwapState = {} },
+    } = useSwapData();
+
+    const { success, hash: swapHash, failReason } = lastSwapState;
+
+    const successData = {
+        amountFrom: `${toSignificantDecimals(amountIn, DECIMALS_TO_SAFE)} ${
+            tokenIn?.onChainFTMetadata?.symbol
+        }`,
+        amountTo: `${toSignificantDecimals(amountOut, DECIMALS_TO_SAFE)} ${
+            tokenOut?.onChainFTMetadata?.symbol
+        }`,
     };
 
     return (
-        <StyledContainer className='buttons-bottom'>
-            <AvatarSuccessIcon />
-            <div
-                className='success-header'
-                data-test-id='swapPageSuccessMessage'
-            >
-                <SafeTranslate id='swap.successTitle' data={messageData} />
-            </div>
-            <div className='buttons-bottom-buttons'>
+        <StyledContainer className="buttons-bottom">
+            {success ? (
+                <>
+                    <AvatarSuccessIcon />
+                    <Header>
+                        <SafeTranslate
+                            id="swap.successTitle"
+                            data={successData}
+                            data-test-id="swapPageSuccessMessage"
+                        />
+                    </Header>
+                </>
+            ) : (
+                <>
+                    <FailIcon />
+                    <Header>
+                        <FailTitle>
+                            <SafeTranslate id="swap.failTitle" />
+                        </FailTitle>
+                        {failReason && (
+                            <Reason className="font-monospace">{failReason}</Reason>
+                        )}
+                    </Header>
+                </>
+            )}
+            <ButtonsWrapper>
                 <FormButton
-                    data-test-id='swapPageContinueAfterSwapButton'
+                    data-test-id="swapPageContinueAfterSwapButton"
                     onClick={onClickContinue}
                 >
-                    <Translate id='button.continue' />
+                    <Translate id="button.continue" />
                 </FormButton>
 
-                {transactionHash && (
-                    <FormButton color='gray-gray' onClick={onClickGoToExplorer}>
-                        <Translate id='button.viewOnExplorer' />
+                {typeof onClickGoToExplorer === 'function' && swapHash && (
```

Comment:
`typeof onClickGoToExplorer === 'function'`
looks scary 😱 not a popular pattern in markup
Could we make it better? Maybe `onClickGoToExplorer  && swapHash  &&` would be enough?

Reply:
Another really simple implementation is using default value of () => {} (empty function)
This way there is no need to check if it is defined and add extra checks.